### ISSUE TITLE
[JIT] Support JAX-style statically shaped nonzero to avoid host-device synchronization

### ISSUE THREAD

**Issue by byronyi** [pytorch/pytorch#62320]
## 🚀 Feature

[`torch.nonzero`](https://pytorch.org/docs/stable/generated/torch.nonzero.html) should accept an optional `size` argument to produce statically shaped output to (1) avoid host-device synchronization and (2) avoid truncating CUDA/LazyTensor IR graph during tracing.

## Motivation

Because the number of nonzero elements is data-dependent, we currently fetch the number of nonzero elements from device, and allocates the corresponding output tensor using the shape information at host runtime. To make the shape information available at host, a host-device synchronization is injected after the fetch of an integer, which stalls the pipeline between Python frontend and device runtime, and also breaks the CUDA/LazyTensor IR graph as we need actual execution to correctly get the concrete shape information and continue the tracing.

## Pitch

Similar to [the nonzero API in JAX](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.nonzero.html), we add a `size` argument to `torch.nonzero`:

> If specified, the first `size` nonzero elements will be returned; if there are fewer nonzero elements than `size` indicates, the index arrays will be zero-padded.

In addition to zero-padding, we could also return a `counts` device tensor to specify the number of nonzero elements at runtime. Similar to what NVIDIA did to enable CUDA graph in [MaskRCNN-ResNet50 MLPerf training benchmark v1.0](https://github.com/mlcommons/training_results_v1.0/blob/8200377f425ae24b6ed6c2816b9273aab0996d43/NVIDIA/benchmarks/maskrcnn/implementations/pytorch/maskrcnn_benchmark/modeling/roi_heads/mask_head/mask_head.py#L35-L47).

## Alternatives

Add a `runtime_shape` tensor field to the tensor class, which will require all ATen operators to be aware of the possible divergence of runtime shape (as tensor of integers on device) and statically inferred shape bound (as `c10::IntArray`).

## Additional context

The following operators either directly call `at::nonzero` or have similar machinery:

```
index.Tensor
masked_select
_unique2
nms
```

PyTorch-XLA encodes the runtime shape information in the opaque tensor handle and [does not truncate IR graph](https://github.com/pytorch/xla/blob/master/docs/source/notes/source_of_recompilation.md#22-what-if-real-dimension-is-queried-on-a-tensor-with-dynamic-shape). However, it is not sound when users inspect `output.shape` in Python which is merely a statically inferred shape bound.

`torch.jit.trace` does shape erasure to support dynamic shape when shape dynamism does not alter the IR graph.

cc @ezyang @gchanan @zou3519 @ngimel @gmagogsfm @ailzhang @asuhan @JackCaoG @mcarilli

**Comment by zou3519** [comment]
cc @Chillee. For vmap, we can't write batching rules for ops that have dynamic behavior like the above so a solution for this would also help vmap.

**Comment by malfet** [comment]
@zou3519 and @jbschlosser to further comment on design/API and how it can benefit some of the workflows

**Comment by zou3519** [comment]
cc @wconstab @bdhirsh do you think this is useful for LazyTensor? (It is a part of the pitch in the issue body). For dynamic upper bounds, instead of using the actual upper bound (which could be pretty high), if a user passes in a `size` argument then there will be a more realistic static size.

**Comment by zou3519** [comment]
Hi-pri for user activity (see https://github.com/pytorch/functorch/issues/256). Unassigning assignees because we are not working on this

**Comment by guangy10** [comment]
I have no plan to work on the GPU impl of the nonzero_static. Just chat with @ezyang , we may need to find the right person who can actually help with this issue.

**Comment by albertz** [comment]
Note, here is my workaround if I really want to avoid the host-device synchronization:

For some given `mask` of shape `[in_len]`:
Instead of:
```python
idx = torch.nonzero(mask).squeeze(1)  # [out_len]
```
Do this trickery:
```python
idx = torch.argsort(mask.to(torch.int8), stable=True, descending=True)  # [in_len]
idx = idx[:out_len]  # [out_len]
```
This avoids the host-device synchronization, but on the other hand, is more complicated, and a bit slower.

**Comment by ngimel** [comment]
`nonzero_static` is implemented

**Comment by laithsakka** [comment]
cc @bobrenjc93

**Comment by ezyang** [comment]
forgot to put it in the autosummary. diff coming

**Comment by vadimkantorov** [comment]
Would it be worth to somehow auto -add stubs in the docs for all public symbols in public namespaces?

At least it would be immediately visible what is forgotten. E.g. also no docs for the `torch.version` and useful `torch.__version__`:
- https://github.com/pytorch/pytorch/issues/32864

**Comment by ezyang** [comment]
The devil is in the details but I'd look positively on an attempt on it; certainly worth checking if claude code could do it. But I also expect to learn there's a problem somehow.
